Cut down conformance checking time by reusing RequirementEnvironments

jrose-apple opened the pull request:
Trade an extra copy for saving the cost of reinitialization when there are a lot of witnesses to check. A small but measurable win when type-checking the standard library.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test compiler performance

xedin reviewed (approved):
Makes sense!

jrose-apple commented:
The failed Linux tests are LLDB tests. Let's try again.

@swift-ci Please test Linux

slavapestov commented:
@jrose-apple Are you sure the requirement environment is the same for each witness? What if one witness is a class method and the other is a protocol extension method? Doesn't the latter have a signature with an extra 'Self' parameter?

jrose-apple commented:
The witness isn't an input to the RequirementEnvironment constructor, so it had better be the same. :-)

slavapestov commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -719,17 +719,19 @@ RequirementMatch swift::matchWitness(TypeChecker &tc,
-      RequirementEnvironment(dc, reqSig, proto, covariantSelf, conformance));
+  if (!cachedReqEnvironment.hasValue()) {
+    cachedReqEnvironment = RequirementEnvironment(dc, reqSig, proto,
+                                                  covariantSelf, conformance);
```
Yeah, so see how we pass 'covariantSelf' to the constructor here. The value of 'covariantSelf' depends on the _witness_; the 'if' statement above only runs if `witness->getDeclContext()->getAsProtocolExtensionContext()`

jrose-apple replied:
…oh shoot. I missed that.

jrose-apple commented:
Okay, this *dramatically* improves type-checking time for some projects, so I guess it's worth figuring out how to do correctly.

slavapestov commented:
@jrose-apple I think the correct way would be to use both `reqSig` and `covariantSelf` as the caching key, with a per-conformance-checker cache. This would even let you reuse requirement environments between requirements, which I think will be a big win if you have a protocol with a bunch of requirements that are all just parametrized as `<Self : P>`.

slavapestov commented:
@jrose-apple _Unfortunately_, I'm afraid the improvement is because something failed to build with the new change. Eg, look at the stats for "AST.NumSourceBuffers | 99,809 | 54,558".

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
This still won't fully succeed but

@swift-ci Please test compiler performance

DougGregor reviewed (approved):
Very nice, LGTM!

jrose-apple commented:
@slavapestov, did you have any remaining comments? (Did I miss anything else? :-) )

jrose-apple commented:
(removed an unused variable)

@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please clean smoke test macOS

slavapestov commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -719,17 +718,23 @@ RequirementMatch swift::matchWitness(TypeChecker &tc,
 
-  Optional<RequirementEnvironment> reqEnvironment(
-      RequirementEnvironment(dc, reqSig, proto, covariantSelf, conformance));
+  WitnessChecker::RequirementEnvironmentCacheKey cacheKey(reqSig,covariantSelf);
```
Maybe extract this out into a getOrCreateRequirementEnvironment(). Also I don't like this trend of omitting spaces to get things to fit in 80 characters. Either live with 81 characters or break the line :)

slavapestov commented on lib/Sema/TypeCheckProtocol.h:
```diff
@@ -430,6 +430,11 @@ struct RequirementMatch {
+public:
+  using RequirementEnvironmentCacheKey =
+      std::pair<const GenericSignature *, const ClassDecl *>;
+  using RequirementEnvironmentCache = llvm::DenseMap<RequirementEnvironmentCacheKey, RequirementEnvironment>;
```
This line is too long

slavapestov commented on lib/Sema/TypeCheckProtocol.h:
```diff
@@ -441,6 +446,8 @@ class WitnessChecker {
   // @_implements(Protocol, DeclName)
   llvm::DenseMap<DeclName, llvm::TinyPtrVector<ValueDecl *>> ImplementsTable;
 
+  RequirementEnvironmentCache ReqEnvironmentCache;
```
Comment?

jrose-apple replied:
I'm not sure there is one other than "cache of requirement environments to be used while checking witnesses", which is pretty vacuous. Am I just being shortsighted, though?

slavapestov reviewed (approved):
Looks good with just a few minor nitpicks.

jrose-apple committed:
Cut down conformance checking time by reusing RequirementEnvironments

Trade an extra copy for saving the cost of reinitialization when there
are a lot of witnesses to check. A small but measurable win when
type-checking the standard library.

Also add a test to make sure we still get the requirement environment
right for covariant 'Self'.

jrose-apple commented:
Hubris to think I could edit without building and get it right first try.

@swift-ci Please smoke test
